Dask setup on Nebari : tutorial

sayantikabanik opened the pull request:
Partially fulfills #39 
Resolves #12 

cc @iameskild @HarshCasper

kcpevey commented on docs/docs/tutorials/dask-nebari-setup.md:
```diff
@@ -0,0 +1,175 @@
+### Conclusion
+
+Kudos ✨, we now have a working dask cluster inside Nebari.
+Excellent to get started with those large datasets.
```
Need to include shutting down the cluster. This is why I'd like to show how we can use a context manager to handle the cluster ([like this](https://gist.github.com/kcpevey/3a9a94ca741decd180c08570a721a9c0)).

Should we show both approaches or just include the context manager here?

iameskild replied:
I think there is some benefit to showing the "manual" approach. Perhaps we could add a note or something towards the bottom that makes a recommendation to use a context manager. 
Something like: 
"Wrapping the dask-gateway in a context manager, like this, is a great way to avoid having to write a lot of boilerplate, with the added benefit that it will ensure the cluster is fully shutdown once it's finished being used."

sayantikabanik replied:
should I add a separate section for the manual approach using `context manager`? 
Or just the recommendation at the end with the `sample code` ?

sayantikabanik replied:
Currently, I have added the `recommendation` with the code snippet. Happy to make any additional tweaks

iameskild replied:
In the same code snippet, could we include the same example from above but being used by this context manager? 

Something like:

```python
with dask_cluster(...) as cluster:
    import dask.array as da
    x = da.random.random((10000, 10000), chunks=(1000, 1000))
    y = x + x.T
    z = y[::2, 5000:].mean(axis=1)
    z.compute()
```

iameskild commented on docs/docs/tutorials/dask-nebari-setup.md:
```diff
@@ -0,0 +1,175 @@
+
+# Introduction to Dask
+
+Working with large datasets could pose computation challenges, running into frequent memory issues is one. Dask is a free and `open-source` library for `parallel computing` in Python, which
```
```suggestion
Working with large datasets can pose a few challenges, running into frequent memory issues is one. [Dask](https://docs.dask.org/en/stable/) is a free and `open-source` library for `parallel computing` in Python,
```

iameskild commented on docs/docs/tutorials/dask-nebari-setup.md:
```diff
@@ -0,0 +1,175 @@
+# Introduction to Dask
+
+Working with large datasets could pose computation challenges, running into frequent memory issues is one. Dask is a free and `open-source` library for `parallel computing` in Python, which
+helps in scaling up the data science workflows leveraging efficient computation
```
```suggestion
enabling data scientist and others to use their favorite PyData tools at scale. 
```

iameskild commented on docs/docs/tutorials/dask-nebari-setup.md:
```diff
@@ -0,0 +1,175 @@
+### Nebari & Dask integration
+
+Nebari uses [dask-gateway](https://gateway.dask.org/) to expose auto-scaling compute clusters automatically configured for the user.
+Dask Gateway provides a secure way to managing dask clusters. We will discuss in detail the pre-defined configuration in later sections of this tutorial.
```
```suggestion
and provides a secure way to managing dask clusters. We will discuss in detail the pre-defined configuration in later sections of this tutorial.
```

iameskild commented on docs/docs/tutorials/dask-nebari-setup.md:
```diff
@@ -0,0 +1,175 @@
+
+### Nebari & Dask integration
+
+Nebari uses [dask-gateway](https://gateway.dask.org/) to expose auto-scaling compute clusters automatically configured for the user.
```
```suggestion
Nebari uses [dask-gateway](https://gateway.dask.org/) to expose auto-scaling compute clusters automatically configured for the user,
```

iameskild commented on docs/docs/tutorials/dask-nebari-setup.md:
```diff
@@ -0,0 +1,175 @@
+
+- `address` : is the rest API that dask-gateway exposes for managing clusters
+- `proxy_address` : is a secure TLS connection to a user defined dask scheduler
+- `auth` is the form of authentication used, which should always be jupyterhub for Nebari
```
```suggestion
- `auth` : is the form of authentication used, which should always be jupyterhub for Nebari
```

iameskild commented on docs/docs/tutorials/dask-nebari-setup.md:
```diff
@@ -0,0 +1,175 @@
+Nebari already has the connection information pre-configured for the user.
+In order to view configuration details run the below command:
+
+- using terminal : `cat /etc/dask/gateway.yaml`
```
```suggestion
- from the terminal : `cat /etc/dask/gateway.yaml`
```

iameskild commented on docs/docs/tutorials/dask-nebari-setup.md:
```diff
@@ -0,0 +1,175 @@
+In order to view configuration details run the below command:
+
+- using terminal : `cat /etc/dask/gateway.yaml`
+- using notebook: `! cat /etc/dask/gateway.yaml`
```
```suggestion
- within a notebook: `! cat /etc/dask/gateway.yaml`
```

iameskild commented on docs/docs/tutorials/dask-nebari-setup.md:
````diff
@@ -0,0 +1,175 @@
+
+```python
+profiles:
+  jupyterlab:
+    - display_name: Small Instance
+      description: Stable environment with 1 cpu / 1 GB ram
+      access: all
+      default: true
+      kubespawner_override:
+        cpu_limit: 1
+        cpu_guarantee: 1
+        mem_limit: 1G
+        mem_guarantee: 1G
+    - display_name: Medium Instance
+      description: Stable environment with 1.5 cpu / 2 GB ram
+      access: yaml
+      groups:
+        - admin
+        - developers
+      users:
+        - bob
+      kubespawner_override:
+        cpu_limit: 1.5
+        cpu_guarantee: 1.25
+        mem_limit: 2G
+        mem_guarantee: 2G
+    - display_name: Large Instance
+      description: Stable environment with 2 cpu / 4 GB ram
+      access: keycloak
+      kubespawner_override:
+        cpu_limit: 2
+        cpu_guarantee: 2
+        mem_limit: 4G
+        mem_guarantee: 4G
````
```suggestion
```

iameskild replied:
I don't think we need to show the jupyterlab profiles here.

sayantikabanik replied:
Maybe having a link could be useful for users who want to dive deeper?

iameskild replied:
Could we pull this whole section out and start a new "How to configure dask gateway profiles" guide instead? See #68.

iameskild replied:
Then, like you said, we can link that document here :)

sayantikabanik replied:
> Could we pull this whole section out and start a new "How to configure dask gateway profiles" guide instead? See #68.

sorry unable to comprehend does by `whole section` mean *Starting a dask cluster on Nebari* sub heading?

iameskild replied:
The selection regarding profiles and the reference to the config yaml.

iameskild commented on docs/docs/tutorials/dask-nebari-setup.md:
````diff
@@ -0,0 +1,175 @@
+options
+```
+
+Once the `ipywidget` shows up we can customise as per our requirement.
````
```suggestion
Once the `ipywidget` shows up, we can customize our cluster.
```

iameskild commented on docs/docs/tutorials/dask-nebari-setup.md:
````diff
@@ -0,0 +1,175 @@
+cluster
+```
+
+We have the option to choose between `manual` scaling and `adaptive` scaling, the benchmarking stats would be
````
Hey @sayantikabanik, what benchmarking stats is this in reference to?

sayantikabanik replied:
I was referring to the user's existing benchmarking stats.
> For example: 

Given any program (DS workflow) comprising heavy components/ETLs have some benchmarking already performed. Max load/Min load, memory, compute etc

iameskild reviewed (changes requested):
This is looking great @sayantikabanik! Thank you! I made some comments and suggestions, let me know what you think. Also, would it be possible to add a screenshot or reference to the dask-labextension?

sayantikabanik commented:
> This is looking great @sayantikabanik! Thank you! I made some comments and suggestions, let me know what you think. Also, would it be possible to add a screenshot or reference to the dask-labextension?

I have added `dask-labextension` section and incorporated the suggestions 🙂

kcpevey reviewed (changes requested):
Thanks @sayantikabanik for this PR! We need to make some significant changes since you've included here both information for the configuration _and_ the tutorial. I'll spend some time separating things out and make it more explicit.

iameskild commented on docs/docs/tutorials/dask-nebari-setup.md:
```diff
@@ -0,0 +1,230 @@
+
+Working with large datasets can pose a few challenges, running into frequent memory issues is one. [Dask](https://docs.dask.org/en/stable/) is a free and `open-source` library for `parallel computing` in Python,
+enabling data scientist and others to use their favorite PyData tools at scale.
+techniques.
```
```suggestion
```

iameskild commented on docs/docs/tutorials/dask-nebari-setup.md:
```diff
@@ -0,0 +1,230 @@
+
+:::note Recommendation
+Wrapping the dask-gateway in a context manager (code below), is a great way to avoid having to write a ton of boilerplate,
+though it comes with an added benefit which ensures the cluster is fully shutdown once the task is complete.
```
```suggestion
and comes with an added benefit that it ensures the cluster is fully shutdown once the task is complete.
```

iameskild reviewed (changes requested):
Hi @sayantikabanik, I made a few more comments. The main one being a request to split out the section about dask profiles (see #68) like @kcpevey mentioned above.

Thanks for your patience :)

iameskild commented on docs/docs/tutorials/dask-nebari-setup.md:
````diff
@@ -0,0 +1,206 @@
+
+Nebari has set of pre-defined options for configuring the dask profiles that we have access to. These can be accessed via Dask Gateway options.
+
+<details>
+<summary> Click to view: Dask profile options for Nebari </summary>
+
+```python
+profiles:
+  dask_worker:
+    "Small Worker":
+      worker_cores_limit: 1
+      worker_cores: 1
+      worker_memory_limit: 1G
+      worker_memory: 1G
+    "Medium Worker":
+      worker_cores_limit: 1.5
+      worker_cores: 1.25
+      worker_memory_limit: 2G
+      worker_memory: 2G
+```
+
+</details>
````
```suggestion
```

iameskild reviewed (approved):
I've removed the reference to profiles and now it should be good to merge!

When we get to #68, we can make an update here that points to the new doc.

Thanks for a lot for this tutorial @sayantikabanik 🎉

sayantikabanik commented:
### TODOs and improvements 

- [ ] Open up all the fold-ups 
- [ ] Write about how to close the dask client from gateway
- [ ] Improve written content on client/gateway/cluster/columnar formats/data persists/ big data side of things 
- [ ] Diagonistic UI plot - explain it more

kcpevey commented:
@sayantikabanik and I had a chat about this. She is going to work on removing the config bits and expanding some of the explanations.

kcpevey commented:
Needs to use NYC Taxi data:
```
import dask.dataframe as dd

ddf = dd.read_parquet(path='s3://nyc-tlc/trip data/yellow_tripdata_2021-01.parquet', storage_options={'anon':True})
ddf
```

sayantikabanik commented:
The dask env on nebari currently doesn't have `s3fs` and `fastparquet or pyarrow` installed to read the NYC taxi dataset. 

- Should we include the `pip install ...` instructions into the tutorial for these two packages or update the dask conda env?

- Below is a tiny example I constructed using the dataset, does this look good?
I can update the related screenshots/descriptions accordingly.

```python
import dask.dataframe as dd

nyc_taxi_dask_df = dd.read_parquet(path='s3://nyc-tlc/trip data/yellow_tripdata_2021-01.parquet', storage_options={'anon':True})
nyc_taxi_dask_df = nyc_taxi_dask_df.persist()
nyc_taxi_dask_df.groupby('payment_type').agg({'fare_amount': ['min', 'max'], 
                                              'tip_amount': ['min', 'max']}).compute()
```

dharhas commented:
Reading parquet files is going to be part of most dask examples so I would say we add `pyarrow`

kcpevey commented:
These have been added to our deployment [here](https://github.com/Quansight/qhub-deployment-quansight/pull/65).

I've opened [an issue](https://github.com/Quansight/qhub/issues/1359) to get them into the default dask environment

sayantikabanik commented:
Can I go ahead with the above code example @dharhas @kcpevey ?

dharhas commented:
@sayantikabanik ideally we should read in 2-3 years of nyc-taxi data i.e. 2019-2022 and then we could show a plot of how the traffic patterns changed due to covid. I tried this for a few minutes before a demo about a month ago and I was having trouble reading in multiple files with wildcards. 

we should be able to read multiple files with 

`ddf = dd.read_parquet(path='s3://nyc-tlc/trip data/yellow_tripdata_2021-*.parquet', storage_options={'anon':True})`

Can you check if that works with Kim's code.

sayantikabanik commented:
I will check this out and try `s3a`

sayantikabanik commented:
Today, I get this error every time I read/list from the s3 bucket. 
`ClientError: An error occurred (AccessDenied) when calling the ListObjectsV2 operation: Access Denied
`

kcpevey commented:
@sayantikabanik I'm getting that same error from using my previous example (which used to work) 
```
ddf = dd.read_parquet(path='s3://nyc-tlc/trip data/yellow_tripdata_2021-01.parquet', storage_options={'anon':True})
```

Do you think we hit it too much and they blocked us? Not sure what else could have changed.

kcpevey commented:
I'm getting the same error from my local machine. That is the same pattern for using dask to access this data that is shown in an example on AWS itself. I think it might be an issue on their end?

sayantikabanik commented:
Can we use the airline dataset instead stored on Nebari GCP?

sayantikabanik commented:
> Can we use the airline dataset instead stored on Nebari GCP? Or Manually download and upload NYC Taxi data + meta info on Nebari cloud and access that?

I have manually downloaded the data for `yellow_taxi` for the years 2021 & 2022 and constructed a parquet dataset folder, @iameskild will be sharing a process guide to upload the data to the public bucket soon.

sayantikabanik commented:
NYC yellow taxi data for the year 2021` is uploaded to the `nebari-public` GCS bucket
Location: `nebari-public/yellow_taxi_NYC_2021`

In order to operate on the dataset, `gcsfs` is required under the `dask` env.
I have opened a [PR](https://github.com/Quansight/qhub-deployment-quansight/pull/71)

sayantikabanik commented:
> 

Locations for NYC taxi dataset from 2019 to 2022
```python
['nebari-public/yellow_taxi_NYC/2019',
 'nebari-public/yellow_taxi_NYC/2020',
 'nebari-public/yellow_taxi_NYC/2021',
 'nebari-public/yellow_taxi_NYC/2022']
```

Detailed path, similar convention is followed for 2020-2022
```python
['nebari-public/yellow_taxi_NYC/2019/01.parquet',
 'nebari-public/yellow_taxi_NYC/2019/02.parquet',
 'nebari-public/yellow_taxi_NYC/2019/03.parquet',
 'nebari-public/yellow_taxi_NYC/2019/04.parquet',
 'nebari-public/yellow_taxi_NYC/2019/05.parquet',
 'nebari-public/yellow_taxi_NYC/2019/06.parquet',
 'nebari-public/yellow_taxi_NYC/2019/07.parquet',
 'nebari-public/yellow_taxi_NYC/2019/08.parquet',
 'nebari-public/yellow_taxi_NYC/2019/09.parquet',
 'nebari-public/yellow_taxi_NYC/2019/10.parquet',
 'nebari-public/yellow_taxi_NYC/2019/11.parquet',
 'nebari-public/yellow_taxi_NYC/2019/12.parquet']
```

sayantikabanik commented:
> 

I am able to read the entire data from 2019-2020 (stored on nebari-public GCS)
```python
import dask.dataframe as dd

nyc_taxi_dask_df = dd.read_parquet(path='gs://nebari-public/yellow_taxi_NYC/*/*', storage_options={'anon':True})
nyc_taxi_dask_df

```

iameskild commented:
@sayantikabanik I tried loading the data and running some basic analysis but I keep running into this error:
```
ArrowInvalid: Parquet magic bytes not found in footer. Either the file is corrupted or this is not a parquet file.
```

I believe I have isolated the issue to the `nebari-public/yellow_taxi_NYC/2020/07.parquet` file. Would you mind trying to upload that file again?

sayantikabanik commented:
> @sayantikabanik I tried loading the data and running some basic analysis but I keep running into this error:
> 
> ```
> ArrowInvalid: Parquet magic bytes not found in footer. Either the file is corrupted or this is not a parquet file.
> ```
> 
> I believe I have isolated the issue to the `nebari-public/yellow_taxi_NYC/2020/07.parquet` file. Would you mind trying to upload that file again?

Thanks @iameskild, could you please recheck?
Below is a basic operation I ran to validate again. 
<img width="1029" alt="Screenshot 2022-08-04 at 7 40 07 AM" src="https://user-images.githubusercontent.com/17350312/182747943-fa41fa5c-781a-4932-b6d4-4fbe3e70797f.png">

kcpevey commented on docs/docs/tutorials/using_dask.md:
```diff
@@ -0,0 +1,234 @@
+
+This corresponds to 32.43GB of data. Running this one-liner would be impossible on most single machines but running this on a dask cluster with 4 works, this can be calculated in under a minute.
+
+...
```
@iameskild can you add whatever goes into this `...` and we can get this merged?

iameskild replied:
Working on this as we speak :)

iameskild commented:
@kcpevey I added some analysis based on @dharhas recommendation above. Once you perform a final review, I believe this is finally ready to be merged :)
